### PR TITLE
Make ColorScheme a WinRT object

### PR DESCRIPTION
**Pull request by carlos-zamora**
## Summary of the Pull Request
ColorScheme is now a WinRT object in the TerminalApp project.

## References
#7141 - ColorScheme is a settings object
#885 - this new settings object will be moved to a new TerminalSettingsModel project

## PR Checklist
* [x] Tests passed

## Detailed Description of the Pull Request / Additional comments
All of the JSON stuff can't be exposed via the idl. So the plan here is that we'll have the TerminalSettingsModel project handle all of the serialization when it's moved over. These functions will be exposed off of the `implementation` namespace, not projected namespace.

## Validation Steps Performed
- [x] Tests passed
- [x] Deployment succeeded

### CODE REVIEW

File: src/cascadia/TerminalApp/ColorScheme.idl [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+        ColorScheme(String name, UInt32 defaultFg, UInt32 defaultBg, UInt32 cursorColor);
+
+        void ApplyScheme(Microsoft.Terminal.TerminalControl.IControlSettings terminalSettings);
+
+        String Name { get; };
+
+        UInt32 Foreground { get; };
+        UInt32 Background { get; };
+        UInt32 SelectionBackground { get; };
+        UInt32 CursorColor { get; };
```
Comment (DHowett) [inline src/cascadia/TerminalApp/ColorScheme.idl:19]: let's not dig this hole deeper. Introduce a Color struct type.

Comment (DHowett) [reply src/cascadia/TerminalApp/ColorScheme.idl:19]: you can specialize its conversion to/from til::color

File: src/cascadia/TerminalApp/lib/TerminalAppLib.vcxproj [hunk 2]
```diff
@@ -263,6 +267,7 @@
     <Midl Include="../CommandKeyChordVisibilityConverter.idl" />
     <Midl Include="../Tab.idl" />
     <Midl Include="../TerminalSettings.idl" />
+    <Midl Include="..\ColorScheme.idl" />
```
Comment (DHowett) [inline src/cascadia/TerminalApp/lib/TerminalAppLib.vcxproj:270]: `/`

File: src/cascadia/TerminalApp/ColorScheme.idl (hunk 1 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/ColorScheme.idl:19]: we probably should have just gone with `Windows.UI.Color` - my bad

File: src/cascadia/TerminalApp/CascadiaSettingsSerialization.cpp [hunk 3]
```diff
@@ -669,9 +670,9 @@ void CascadiaSettings::_LayerOrCreateColorScheme(const Json::Value& schemeJson)
 // Return Value:
 // - a ColorScheme that can be layered with the given json object, iff such a
 //   color scheme exists.
-ColorScheme* CascadiaSettings::_FindMatchingColorScheme(const Json::Value& schemeJson)
+implementation::ColorScheme* CascadiaSettings::_FindMatchingColorScheme(const Json::Value& schemeJson)
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/CascadiaSettingsSerialization.cpp:673]: oh hey this should probably not be a bare `implementation::ColorScheme*` anymore. It should probably be the `com_ptr<implementation::ColorScheme>` now, right?

File: src/cascadia/TerminalApp/CascadiaSettingsSerialization.cpp [hunk 4]
```diff
@@ -681,7 +682,7 @@ ColorScheme* CascadiaSettings::_FindMatchingColorScheme(const Json::Value& schem
             // maybe not the _safest_ thing, but we have a mind to make Profile
             // and ColorScheme winrt types in the future, so this will be safer
             // then.
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/CascadiaSettingsSerialization.cpp:684]: And hey yank this comment

File: src/cascadia/TerminalApp/ColorScheme.idl (hunk 1 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalApp/ColorScheme.idl:19]: Updated ColorScheme to use Windows.UI.Color. Should I go ahead and propagate that change to `ICoreSettings` too? I'm already here.

File: src/cascadia/TerminalApp/lib/TerminalAppLib.vcxproj (hunk 2 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalApp/lib/TerminalAppLib.vcxproj:270]: haha. I'm gonna be hitting this _a lot_ because VS adds it in like this. I'll keep an eye out for it :P

File: src/cascadia/ut_app/JsonTests.cpp [hunk 5]
```diff
@@ -94,10 +94,10 @@ namespace TerminalAppUnitTests
...
         auto scheme = implementation::ColorScheme::FromJson(schemeObject);
         VERIFY_ARE_EQUAL(L"Campbell", scheme->Name());
-        VERIFY_ARE_EQUAL(til::color{ARGB(0, 0xf2, 0xf2, 0xf2)}, scheme->Foreground());
-        VERIFY_ARE_EQUAL(til::color{ARGB(0, 0x0c, 0x0c, 0x0c)}, scheme->Background());
-        VERIFY_ARE_EQUAL(til::color{ARGB(0, 0x13, 0x13, 0x13)}, scheme->SelectionBackground());
-        VERIFY_ARE_EQUAL(til::color{ARGB(0, 0xFF, 0xFF, 0xFF)}, scheme->CursorColor());
+        VERIFY_ARE_EQUAL(til::color{ ARGB(0, 0xf2, 0xf2, 0xf2) }, scheme->Foreground());
+        VERIFY_ARE_EQUAL(til::color{ ARGB(0, 0x0c, 0x0c, 0x0c) }, scheme->Background());
+        VERIFY_ARE_EQUAL(til::color{ ARGB(0, 0x13, 0x13, 0x13) }, scheme->SelectionBackground());
+        VERIFY_ARE_EQUAL(til::color{ ARGB(0, 0xFF, 0xFF, 0xFF) }, scheme->CursorColor());
```
Comment (DHowett) [inline src/cascadia/ut_app/JsonTests.cpp:100]: til::color already takes argb as its normal constructor params, is the bit where we go from A R G B -> COLORREF necessary?

File: src/cascadia/ut_app/JsonTests.cpp [hunk 6]
```diff
@@ -107,7 +107,7 @@ namespace TerminalAppUnitTests
         for (size_t i = 0; i < expectedCampbellTable.size(); i++)
         {
             const auto& expected = expectedCampbellTable.at(i);
-            const auto& actual = scheme.GetTable().at(i);
+            const auto& actual = til::color{ scheme->Table().at(static_cast<uint32_t>(i)) };
```
Comment (DHowett) [inline src/cascadia/ut_app/JsonTests.cpp:110]: this is invalid -- you've captured a reference to a local temporary.

you want this:

```suggestion
            const til::color actual{ scheme->Table().at(static_cast<uint32_t>(i)) };
```

File: src/cascadia/TerminalApp/CascadiaSettings.h [hunk 7]
```diff
@@ -78,7 +80,7 @@ class TerminalApp::CascadiaSettings final
     static std::filesystem::path GetDefaultSettingsPath();
 
     const Profile* FindProfile(GUID profileGuid) const noexcept;
-    const ColorScheme* GetColorSchemeForProfile(const GUID profileGuid) const;
+    const winrt::TerminalApp::ColorScheme* GetColorSchemeForProfile(const GUID profileGuid) const;
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/CascadiaSettings.h:83]: This should probably just be a `winrt::TerminalApp::ColorScheme` right? I'm pretty sure that pointers to the projected type are a no-no (@DHowett can correct me if I'm wrong)

Comment (DHowett) [reply src/cascadia/TerminalApp/CascadiaSettings.h:83]: Absolutely a no-no

File: src/cascadia/TerminalApp/ColorScheme.cpp [hunk 8]
```diff
@@ -64,20 +67,20 @@ ColorScheme::~ColorScheme()
 // - Apply our values to the given TerminalSettings object. Sets the foreground,
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/ColorScheme.cpp:67]: update comment

File: src/cascadia/TerminalApp/ColorScheme.h [hunk 9]
```diff
@@ -18,48 +18,53 @@ Author(s):
 #include "TerminalSettings.h"
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/ColorScheme.h:18]: Don't need this anymore

**Comment by DHowett** [comment]
I don't get it. All that 7a3ca00a3df630a50802cd177a89c739803e3d8b did was rename some variables.

The responsibility to apply a color scheme to an IControlSettings lies inside TerminalApp where TerminalSettings lives. You haven't done anything to break the dependency from colorscheme to terminal **control** (where IControlSettings now lives)

**Comment by carlos-zamora** [comment]
> I don't get it. All that [7a3ca00](https://github.com/microsoft/terminal/commit/7a3ca00a3df630a50802cd177a89c739803e3d8b) did was rename some variables.
> 
> The responsibility to apply a color scheme to an IControlSettings lies inside TerminalApp where TerminalSettings lives. You haven't done anything to break the dependency from colorscheme to terminal **control** (where IControlSettings now lives)

At some point I had ColorScheme.idl referencing a TerminalSettings. That shouldn't be allowed. Making it reference an IControlSettings is _better_, but you're right, all of that logic should be moved to TermApp.

I was thinking of this little change being a stepping stone to #885's last step (Move the new WinRT settings objects into TerminalSettings project). We don't _need_ the separation until then, at which point I'll also move the similar logic over for Profile.

TLDR: referencing IControlSettings > TerminalSettings

**Comment by DHowett** [comment]
I actually disagree that it’s better!

Eventually, the interfaces for settings will become read-only and you will need to transition this code _back_ to using a class type that has setters on it. Moving to the interface right now is going to be a roadblock for when you inevitably have to concentrate the logic in App.

**Comment by DHowett** [comment]
(cf #7219)

File: src/cascadia/TerminalApp/ColorScheme.cpp [hunk 10]
```diff
@@ -138,32 +141,37 @@ void ColorScheme::LayerJson(const Json::Value& json)
     }
 }
 
-std::wstring_view ColorScheme::GetName() const noexcept
+winrt::hstring ColorScheme::Name() const noexcept
 {
-    return { _schemeName };
+    return { _schemeName.c_str() };
```
Comment (DHowett) [inline src/cascadia/TerminalApp/ColorScheme.cpp:146]: hol up, what

Comment (DHowett) [reply src/cascadia/TerminalApp/ColorScheme.cpp:146]: don't do this. if you're ever going from one string type to another string type, _don't use the data pointer unless absolutely necessary_.

winrt::hstring has a constructor that just takes a wstring. just use that one.

File: src/cascadia/TerminalApp/ColorScheme.cpp [hunk 11]
```diff
@@ -138,32 +141,37 @@ void ColorScheme::LayerJson(const Json::Value& json)
...
 
-std::wstring_view ColorScheme::GetName() const noexcept
+winrt::hstring ColorScheme::Name() const noexcept
 {
-    return { _schemeName };
+    return { _schemeName.c_str() };
 }
 
-std::array<til::color, COLOR_TABLE_SIZE>& ColorScheme::GetTable() noexcept
+winrt::com_array<Color> ColorScheme::Table() noexcept
```
Comment (DHowett) [inline src/cascadia/TerminalApp/ColorScheme.cpp:149]: it is critically important that we validate the users of this function.
The old version _let recipients edit the color table_
This version _does not_

File: src/cascadia/TerminalApp/ColorScheme.cpp [hunk 12]
```diff
@@ -138,32 +141,37 @@ void ColorScheme::LayerJson(const Json::Value& json)
...
 {
-    return _table;
+    winrt::com_array<Color> result = winrt::com_array<Color>(COLOR_TABLE_SIZE);
```
Comment (DHowett) [inline src/cascadia/TerminalApp/ColorScheme.cpp:151]: unexpected/unnecessary constructor call

```suggestion
    winrt::com_array<Color> result{ COLOR_TABLE_SIZE };
```

File: src/cascadia/TerminalApp/ColorScheme.cpp [hunk 13]
```diff
@@ -138,32 +141,37 @@ void ColorScheme::LayerJson(const Json::Value& json)
...
+    for (uint32_t i = 0; i < _table.size(); i++)
```
Comment (DHowett) [inline src/cascadia/TerminalApp/ColorScheme.cpp:152]: can we std::transform or std::foreach this?

File: src/cascadia/TerminalApp/ColorScheme.idl [hunk 14]
```diff
@@ -0,0 +1,23 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import "../TerminalSettings.idl";
```
Comment (DHowett) [inline src/cascadia/TerminalApp/ColorScheme.idl:4]: you don't need this. you're using `IControlSettings` below, not `TerminalSettings`.

File: src/cascadia/TerminalApp/GlobalAppSettings.cpp [hunk 15]
```diff
@@ -213,7 +213,7 @@ void GlobalAppSettings::LayerJson(const Json::Value& json)
 // - <none>
 void GlobalAppSettings::AddColorScheme(ColorScheme scheme)
 {
-    std::wstring name{ scheme.GetName() };
+    std::wstring name{ scheme.Name() };
```
Comment (DHowett) [inline src/cascadia/TerminalApp/GlobalAppSettings.cpp:216]: _i hate that we're going from wstring (inside scheme) -> hstring -> wstring -> hstring again_

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 16]
```diff
@@ -634,10 +634,10 @@ namespace winrt::TerminalApp::implementation
                                         newTerminalArgs.Profile().empty());
 
         // Lookup the name of the color scheme used by this profile.
-        const auto* scheme = _settings->GetColorSchemeForProfile(profileGuid);
+        const auto scheme = _settings->GetColorSchemeForProfile(profileGuid);
         // If they explicitly specified `null` as the scheme (indicating _no_ scheme), log
         // that as the empty string.
-        const auto schemeName = scheme ? scheme->GetName() : L"\0";
+        const auto schemeName = scheme ? scheme.Name() : L"\0";
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TerminalPage.cpp:640]: surprised this compiles

File: src/cascadia/TerminalApp/ColorScheme.cpp (hunk 11 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalApp/ColorScheme.cpp:149]: Looks like every recipient only reads these values. So that's good. I'll add a const to the function header to make that clear though.

The closest I found was that `TerminalSettings` has a `SetColorTableEntry()`, but it makes sense that we don't need that (right now, at least) because we're only reading the settings, not writing.

File: src/cascadia/TerminalApp/GlobalAppSettings.cpp (hunk 15 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalApp/GlobalAppSettings.cpp:216]: _yeaaaaah_. Once every settings object is a winrt type, everything will just be an hstring all the way through.

I should've checked this actually, but does `JsonUtils::GetValueForKey()` work for an hstring?

Comment (DHowett) [reply src/cascadia/TerminalApp/GlobalAppSettings.cpp:216]: Yes.

File: src/cascadia/TerminalApp/ColorScheme.cpp [hunk 17]
```diff
@@ -143,16 +143,13 @@ void ColorScheme::LayerJson(const Json::Value& json)
...
-winrt::com_array<Color> ColorScheme::Table() noexcept
+winrt::com_array<Color> ColorScheme::Table() const noexcept
 {
-    winrt::com_array<Color> result = winrt::com_array<Color>(COLOR_TABLE_SIZE);
-    for (uint32_t i = 0; i < _table.size(); i++)
-    {
-        result[i] = _table[i];
-    }
+    winrt::com_array<Color> result{ COLOR_TABLE_SIZE };
+    std::transform(_table.begin(), _table.end(), result.begin(), [](til::color c) -> Color { return c; });
```
Comment (DHowett) [inline src/cascadia/TerminalApp/ColorScheme.cpp:152]: this is fine because a til::color is small enough to fit in a single register (or two registers), but it could be an expensive copy for any other struct. prefer `const x&` unless otherwise okay
